### PR TITLE
[WIP] feat(events): provide the ability to specify key for global events

### LINKED ISSUE #7308: [Feature request] Global key event filtering
When having a component that wants to listen to global key events, using `(body:keyup.escape)` crashes with an `EXCEPTION: not implemented`, whereas `(body:keyup)` works fine. Is this actually being worked on, or just not in the roadmap at all?

Here is my use case: I'm trying to build a modal component, and I want to dismiss it when the user presses escape. And any listener tied to the modal itself never triggers, so I need to listen to the keyup on `body`.

``` typescript
@Component({
    selector: "my-modal",
    host: {
        "(body:keyup.escape)": "dismiss()"
    }
    ...
})
export class MyModal {
    @Input("MyModalOpen") open = false;
    @Output("MyModalOpenChange") closed = new EventEmitter<boolean>();
    ...
}
```

The only solution I see right now is catch every keyup and filter myself, which feels dirty. I'm open to any alternative solutions. :smile: 
Thanks for your help.

### PR DESCRIPTION
**Pull request by DzmitryShylovich**
**What kind of change does this PR introduce?** (check one with "x")

```
[ ] Bugfix
[x] Feature
[x] Code style update (formatting, local variables)
[x] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Other... Please describe:
```

This PR adds support for:
-  Event keys on global events:

**Before**

```
 @HostListener('document:keyup') onKeyUp() { ... }
```

**After**

```
 @HostListener('document:keyup.esc') onEscKey() { ... }
```
- Global hammerjs events:

**Before**

```
@HostListener('swiperight', ['$event']) handleSwipeRight(event) { ... }
```

**After**

```
@HostListener('document:swiperight', ['$event']) handleSwipeRight(event) { ... }
```

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

Fixes https://github.com/angular/angular/issues/7308 and https://github.com/angular/angular/issues/12566

### CODE REVIEW

**Comment by DzmitryShylovich** [comment]
I signed it!

**Comment by DzmitryShylovich** [comment]
I signed it!

File: modules/@angular/platform-browser/src/dom/events/event_manager.ts [hunk 1]
```diff
@@ -61,10 +58,10 @@ export class EventManagerPlugin {
   supports(eventName: string): boolean { return false; }
 
   addEventListener(element: HTMLElement, eventName: string, handler: Function): Function {
```
Comment (vicb) [inline modules/@angular/platform-browser/src/dom/events/event_manager.ts]: please rather make the class and those 2 methods abstract

File: modules/@angular/platform-browser/src/dom/events/hammer_gestures.ts [hunk 2]
```diff
@@ -56,27 +55,30 @@ export class HammerGesturesPlugin extends HammerGesturesPluginCommon {
...
     return zone.runOutsideAngular(() => {
       // Creating the manager bind events, must be done outside of angular
-      var mc = this._config.buildHammer(element);
-      var callback = function(eventObj: any /** TODO #???? */) {
-        zone.runGuarded(function() { handler(eventObj); });
-      };
-      mc.on(eventName, callback);
-      return () => { mc.off(eventName, callback); };
+      const mc: HammerInstance = this._config.buildHammer(element);
+      const callback = (eventObj: Event) => { zone.runGuarded(() => { handler(eventObj); }); };
```
Comment (vicb) [inline modules/@angular/platform-browser/src/dom/events/hammer_gestures.ts]: event

File: modules/@angular/platform-browser/src/dom/events/key_events.ts [hunk 3]
```diff
@@ -7,22 +7,23 @@
...
-var modifierKeyGetters: {[key: string]: (event: KeyboardEvent) => boolean} = {
+const MODIFIER_KEYS = ['alt', 'control', 'meta', 'shift'];
+const MODIFIER_KEY_GETTERS: {[key: string]: (event: KeyboardEvent) => boolean} = {
   'alt': (event: KeyboardEvent) => event.altKey,
   'control': (event: KeyboardEvent) => event.ctrlKey,
   'meta': (event: KeyboardEvent) => event.metaKey,
   'shift': (event: KeyboardEvent) => event.shiftKey
 };
 
+export interface ParsedEvent {
```
Comment (vicb) [inline modules/@angular/platform-browser/src/dom/events/key_events.ts]: do this need to be exported ?

File: modules/@angular/platform-browser/src/dom/events/key_events.ts [hunk 4]
```diff
@@ -99,17 +95,17 @@ export class KeyEventsPlugin extends EventManagerPlugin {
     return fullKey;
   }
 
-  static eventCallback(element: HTMLElement, fullKey: any, handler: Function, zone: NgZone):
-      Function {
-    return (event: any /** TODO #9100 */) => {
-      if (StringWrapper.equals(KeyEventsPlugin.getEventFullKey(event), fullKey)) {
+  /** @internal */
```
Comment (vicb) [inline modules/@angular/platform-browser/src/dom/events/key_events.ts]: no need on private

File: modules/@angular/platform-browser/src/dom/events/key_events.ts [hunk 5]
```diff
@@ -99,17 +95,17 @@ export class KeyEventsPlugin extends EventManagerPlugin {
...
+  private static eventCallback(fullKey: string, handler: Function, zone: NgZone): Function {
+    return (event: KeyboardEvent) => {
+      if (KeyEventsPlugin.getEventFullKey(event) === fullKey) {
         zone.runGuarded(() => handler(event));
       }
     };
   }
 
   /** @internal */
```
Comment (vicb) [inline modules/@angular/platform-browser/src/dom/events/key_events.ts]: same thing

File: modules/@angular/platform-browser/src/dom/events/key_events.ts (hunk 4 above)
Comment (vicb) [inline modules/@angular/platform-browser/src/dom/events/key_events.ts]: rename to _eventCallback for consistency

**Comment by vicb** [comment]
Please rebase on master.

(While the best practice is indeed to merge master into feature branches we do usually rebase only we really have to share the dev)

**Comment by vicb** [comment]
Thanks for your PR, could you please fix the description (before/after) and a test.

**Comment by vicb** [comment]
isn't `@HostListener` in contradiction with `document:` (semantic) ?

**Comment by zoechi** [comment]
> isn't @HostListener in contradiction with document: (semantic) ?

@vicb I think so, but that's not related to this PR, is it? 
This is the same for any declarative global event listener, also for

    <div (window:resize)="onResize($event)">

which is also not related to the `<div>` in any way.

File: modules/@angular/platform-browser/src/dom/events/key_events.ts (hunk 3 above)
Comment (DzmitryShylovich) [inline modules/@angular/platform-browser/src/dom/events/key_events.ts]: it's required in test file https://github.com/DzmitryShylovich/angular/blob/c05a92d947fe7cc677e6db14f0294673ba1ea366/modules/%40angular/platform-browser/test/dom/events/key_events_spec.ts
(it's not actually imported here but if I don't export ParsedEvent ts will throw errors)

**Comment by DzmitryShylovich** [comment]
Rebased, squashed and addressed all of your comments @vicb

File: modules/@angular/platform-browser/src/dom/events/dom_events.ts [hunk 6]
```diff
@@ -18,17 +17,13 @@ export class DomEventsPlugin extends EventManagerPlugin {
   supports(eventName: string): boolean { return true; }
 
   addEventListener(element: HTMLElement, eventName: string, handler: Function): Function {
-    var zone = this.manager.getZone();
-    var outsideHandler = (event: any /** TODO #9100 */) => zone.runGuarded(() => handler(event));
-    return this.manager.getZone().runOutsideAngular(
-        () => getDOM().onAndCancel(element, eventName, outsideHandler));
+    const zone: NgZone = this.manager.getZone();
+    const outsideHandler = (event: Event) => zone.runGuarded(() => handler(event));
```
Comment (vicb) [inline modules/@angular/platform-browser/src/dom/events/dom_events.ts]: don't think `event: Event` is right for angular events

File: modules/@angular/platform-browser/src/dom/events/event_manager.ts [hunk 7]
```diff
@@ -51,17 +50,14 @@ export class EventManager {
...
-  // That is equivalent to having supporting $event.target
-  supports(eventName: string): boolean { return false; }
+  abstract supports(eventName: string): boolean;
 
-  addEventListener(element: HTMLElement, eventName: string, handler: Function): Function {
-    throw 'not implemented';
-  }
+  abstract addEventListener(element: HTMLElement, eventName: string, handler: Function): Function;
 
   addGlobalEventListener(element: string, eventName: string, handler: Function): Function {
```
Comment (vicb) [inline modules/@angular/platform-browser/src/dom/events/event_manager.ts]: make it abstract. remove impl

File: modules/@angular/platform-browser/src/dom/events/hammer_gestures.ts [hunk 8]
```diff
@@ -55,27 +54,30 @@ export class HammerGesturesPlugin extends HammerGesturesPluginCommon {
...
     return zone.runOutsideAngular(() => {
       // Creating the manager bind events, must be done outside of angular
-      var mc = this._config.buildHammer(element);
-      var callback = function(eventObj: any /** TODO #???? */) {
-        zone.runGuarded(function() { handler(eventObj); });
-      };
-      mc.on(eventName, callback);
-      return () => { mc.off(eventName, callback); };
+      const mc: HammerInstance = this._config.buildHammer(element);
+      const callback = (event: Event) => { zone.runGuarded(() => { handler(event); }); };
```
Comment (vicb) [inline modules/@angular/platform-browser/src/dom/events/hammer_gestures.ts]: is `event: Event` true with hammer ?

**Review by vicb — CHANGES_REQUESTED** [review]
Please add test for the use cases described in the PR comment

**Pull request closed without merging**